File: minigit-4/ts/src/commands/branch.ts
// branch command - List, create, or delete branches

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, isValidBranchName, shortSha } from '../utils';
import {
  getBranches,
  getCurrentBranch,
  getHeadCommit,
  branchExists,
  updateBranch,
  deleteBranch,
  resolveRevision,
  resolveRef,
} from '../refs';
import { readObject, parseCommitContent, objectExists } from '../objects';

export function branch(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let deleteFlag = false;
  let forceDelete = false;
  let rename = false;
  let verbose = false;
  const positional: string[] = [];

  for (const arg of args) {
    if (arg === '-d') {
      deleteFlag = true;
    } else if (arg === '-D') {
      forceDelete = true;
      deleteFlag = true;
    } else if (arg === '-m') {
      rename = true;
    } else if (arg === '-v' || arg === '--verbose') {
      verbose = true;
    } else if (!arg.startsWith('-')) {
      positional.push(arg);
    }
  }

  if (deleteFlag) {
    // Delete branch
    if (positional.length === 0) {
      console.error('fatal: branch name required');
      return 1;
    }

    const branchName = positional[0];
    return deleteBranchCmd(repoRoot, branchName, forceDelete);
  }

  if (rename) {
    // Rename branch
    if (positional.length < 2) {
      console.error('fatal: need old and new branch names');
      return 1;
    }

    const oldName = positional[0];
    const newName = positional[1];
    return renameBranch(repoRoot, oldName, newName);
  }

  if (positional.length === 0) {
    // List branches
    return listBranches(repoRoot, verbose);
  }

  // Create branch
  const branchName = positional[0];
  const startPoint = positional[1];
  return createBranch(repoRoot, branchName, startPoint);
}

function listBranches(repoRoot: string, verbose: boolean): number {
  const branches = getBranches(repoRoot);
  const currentBranch = getCurrentBranch(repoRoot);

  for (const branch of branches) {
    const isCurrent = branch === currentBranch;
    const prefix = isCurrent ? '* ' : '  ';

    if (verbose) {
      const sha = resolveRef(repoRoot, branch);
      if (sha) {
        let message = '';
        try {
          const { content } = readObject(repoRoot, sha);
          const info = parseCommitContent(content);
          message = info.message.split('\n')[0];
        } catch {
          // Ignore
        }
        console.log(`${prefix}${branch} ${shortSha(sha)} ${message}`);
      } else {
        console.log(`${prefix}${branch}`);
      }
    } else {
      console.log(`${prefix}${branch}`);
    }
  }

  return 0;
}

function createBranch(repoRoot: string, branchName: string, startPoint?: string): number {
  // Validate branch name
  if (!isValidBranchName(branchName)) {
    console.error(`fatal: '${branchName}' is not a valid branch name`);
    return 1;
  }

  // Check if already exists
  if (branchExists(repoRoot, branchName)) {
    console.error(`fatal: branch '${branchName}' already exists`);
    return 1;
  }

  // Get starting commit
  let sha: string | null;
  if (startPoint) {
    sha = resolveRevision(repoRoot, startPoint);
    if (!sha) {
      console.error(`fatal: not a valid object name: '${startPoint}'`);
      return 1;
    }
  } else {
    sha = getHeadCommit(repoRoot);
    if (!sha) {
      console.error('fatal: not a valid object name: HEAD');
      return 1;
    }
  }

  updateBranch(repoRoot, branchName, sha);
  return 0;
}

function deleteBranchCmd(repoRoot: string, branchName: string, force: boolean): number {
  // Check if branch exists
  if (!branchExists(repoRoot, branchName)) {
    console.error(`error: branch '${branchName}' not found`);
    return 1;
  }

  // Check if trying to delete current branch
  const currentBranch = getCurrentBranch(repoRoot);
  if (branchName === currentBranch) {
    console.error(`error: Cannot delete branch '${branchName}' checked out`);
    return 1;
  }

  // If not force, check if branch is merged
  if (!force) {
    const branchSha = resolveRef(repoRoot, branchName);
    const headSha = getHeadCommit(repoRoot);

    if (branchSha && headSha && !isAncestor(repoRoot, branchSha, headSha)) {
      console.error(`error: branch '${branchName}' is not fully merged`);
      console.error(`If you are sure you want to delete it, run 'minigit branch -D ${branchName}'`);
      return 1;
    }
  }

  deleteBranch(repoRoot, branchName);
  console.log(`Deleted branch ${branchName}`);
  return 0;
}

function renameBranch(repoRoot: string, oldName: string, newName: string): number {
  // Validate new name
  if (!isValidBranchName(newName)) {
    console.error(`fatal: '${newName}' is not a valid branch name`);
    return 1;
  }

  // Check if old branch exists
  if (!branchExists(repoRoot, oldName)) {
    console.error(`error: branch '${oldName}' not found`);
    return 1;
  }

  // Check if new name already exists
  if (branchExists(repoRoot, newName)) {
    console.error(`fatal: branch '${newName}' already exists`);
    return 1;
  }

  // Get the SHA of the old branch
  const sha = resolveRef(repoRoot, oldName);
  if (!sha) {
    console.error(`error: branch '${oldName}' not found`);
    return 1;
  }

  // Create new branch
  updateBranch(repoRoot, newName, sha);

  // Delete old branch
  deleteBranch(repoRoot, oldName);

  // Update HEAD if it was pointing to old branch
  const currentBranch = getCurrentBranch(repoRoot);
  if (currentBranch === oldName) {
    const headPath = path.join(repoRoot, '.minigit', 'HEAD');
    fs.writeFileSync(headPath, `ref: refs/heads/${newName}\n`);
  }

  return 0;
}

function isAncestor(repoRoot: string, commit: string, descendant: string): boolean {
  // BFS to check if commit is an ancestor of descendant
  const visited = new Set<string>();
  const queue = [descendant];

  while (queue.length > 0) {
    const current = queue.shift()!;

    if (current === commit) {
      return true;
    }

    if (visited.has(current)) {
      continue;
    }
    visited.add(current);

    if (!objectExists(repoRoot, current)) {
      continue;
    }

    try {
      const { type, content } = readObject(repoRoot, current);
      if (type !== 'commit') continue;

      const info = parseCommitContent(content);
      queue.push(...info.parents);
    } catch {
      // Ignore errors
    }
  }

  return false;
}

File: minigit-4/py/commands/branch.py
"""branch command - List, create, or delete branches."""

import sys
import re
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import read_object, Commit, resolve_object
from refs import (
    list_branches, get_current_branch, create_branch, delete_branch,
    rename_branch, get_ref, is_branch_merged, resolve_head
)


def validate_branch_name(name: str) -> bool:
    """Check if branch name is valid."""
    if not name:
        return False
    if name.startswith('-'):
        return False
    if ' ' in name:
        return False
    if '..' in name:
        return False
    if name.startswith('.') or name.endswith('.'):
        return False
    if '@{' in name:
        return False
    # Check for invalid characters
    invalid_chars = ['~', '^', ':', '\\', '?', '*', '[']
    for c in invalid_chars:
        if c in name:
            return False
    return True


def run(args):
    """Manage branches."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    delete = False
    force_delete = False
    rename = False
    verbose = False
    branch_name = None
    start_point = None
    old_name = None
    new_name = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-d':
            delete = True
        elif arg == '-D':
            force_delete = True
            delete = True
        elif arg == '-m':
            rename = True
        elif arg == '-v':
            verbose = True
        elif not arg.startswith('-'):
            if rename:
                if old_name is None:
                    old_name = arg
                else:
                    new_name = arg
            elif delete:
                branch_name = arg
            elif branch_name is None:
                branch_name = arg
            else:
                start_point = arg
        i += 1

    # Handle rename
    if rename:
        if not old_name or not new_name:
            print("error: branch rename requires old and new names", file=sys.stderr)
            return 1

        if not validate_branch_name(new_name):
            print(f"fatal: '{new_name}' is not a valid branch name", file=sys.stderr)
            return 1

        if rename_branch(old_name, new_name, repo_root):
            return 0
        else:
            print(f"error: failed to rename branch '{old_name}'", file=sys.stderr)
            return 1

    # Handle delete
    if delete:
        if not branch_name:
            print("error: branch name required", file=sys.stderr)
            return 1

        current = get_current_branch(repo_root)
        if branch_name == current:
            print(f"error: Cannot delete branch '{branch_name}' checked out", file=sys.stderr)
            return 1

        # Check if branch exists
        if get_ref(f'refs/heads/{branch_name}', repo_root) is None:
            print(f"error: branch '{branch_name}' not found", file=sys.stderr)
            return 1

        # Check if merged (unless force delete)
        if not force_delete:
            if current and not is_branch_merged(branch_name, current, repo_root):
                print(f"error: branch '{branch_name}' is not fully merged", file=sys.stderr)
                print("If you are sure you want to delete it, run 'minigit branch -D {}'".format(branch_name), file=sys.stderr)
                return 1

        if delete_branch(branch_name, repo_root):
            print(f"Deleted branch {branch_name}")
            return 0
        else:
            print(f"error: failed to delete branch '{branch_name}'", file=sys.stderr)
            return 1

    # Handle create
    if branch_name:
        if not validate_branch_name(branch_name):
            print(f"fatal: '{branch_name}' is not a valid branch name", file=sys.stderr)
            return 1

        # Get start point
        if start_point:
            try:
                sha = resolve_object(start_point, repo_root)
            except ValueError as e:
                print(f"fatal: {e}", file=sys.stderr)
                return 1
        else:
            sha = resolve_head(repo_root)
            if not sha:
                print("fatal: not a valid object name: 'HEAD'", file=sys.stderr)
                return 1

        if create_branch(branch_name, sha, repo_root):
            return 0
        else:
            print(f"fatal: branch '{branch_name}' already exists", file=sys.stderr)
            return 1

    # List branches
    branches = list_branches(repo_root)
    current = get_current_branch(repo_root)

    for branch in branches:
        if branch == current:
            prefix = '* '
        else:
            prefix = '  '

        if verbose:
            sha = get_ref(f'refs/heads/{branch}', repo_root)
            if sha:
                try:
                    _, data = read_object(sha, repo_root)
                    commit = Commit.deserialize(data)
                    msg = commit.message.split('\n')[0][:50]
                    print(f'{prefix}{branch} {sha[:7]} {msg}')
                except:
                    print(f'{prefix}{branch} {sha[:7]}')
            else:
                print(f'{prefix}{branch}')
        else:
            print(f'{prefix}{branch}')

    return 0
